keps/sig-api-machinery/customresources-pruning-and-defaulting.md

```diff
@@ -0,0 +1,486 @@
+> 
+> The schema `s` above means to either have a `map[string]int64` or a `struct` with `a` of type string. This is a special case of polymorphism because for the same JSON path is typed with two different types in the Golang sense (struct and `map[string]T`), but same types in the JSON sense (JSON object).
+
+Hence, we have to add the following restriction to OpenAPI validation schemata:
+
+**Restriction 1 on Object Polymorphism:** reject CRD OpenAPI validation schemata `s` with `skel(s)` having `properties` and one of  `additionalProperties` or `patternProperties` being defined for the same JSON path.
+
+### Defaulting
+
+**Restriction 2:** Defaults may only be defined outside of `anyOf`, `allOf`, `oneOf` and `not`. All other defaults are rejected in CRD validation.
+
+**Consequence:** defaults are not conditional, but static. I.e. they don’t depend on the validity of other parts of the input object.
+
+#### Example 11
+ 
+> In a OpenAPI validation schema, one can write: 
+>
+>```json
+>{
+>  "anyOf": [
+>    { "properties": {"a": {"pattern": "1"}, "b": {"default": "1"}}},
+>    { "properties": {"a": {"pattern": "2"}, "b": {"default": "2"}}}
+>  ]
+>}
+>``` 
+>
+> This says to default `b` to the same value as `a`. This cannot be expressed with a skeleton schema. In this sense, the defaults are static.
+
+Pruning and defaulting will be applied after every unstructured JSON decoding and before conversion.
+
+### Excluding values from Pruning
+
+There are cases where parts of an object are verbatim JSON, i.e. without any applied schema and especially without a complete specification which allows to apply pruning. E.g. the [`JSONSchemaProps` types in apiextensions-apiserver contain fields of type JSON](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types_jsonschema.go#L28) for defaults, examples and enums. These should never be pruned (and they are not pruned while decoding using JSON marshalling).
+
+Hence, we need a mechanism to express that in the OpenAPI validation schema (compare https://github.com/kubernetes/kubernetes/pull/64558#issuecomment-403564033).
+
+**Raw JSON Option 1:** add a format `json`, e.g.:
+
+```json
+{"properties": {"x": {"format": "json"}}}
+```
+
+In this example "x" is excluded from pruning. Note that you can still use any kind of OpenAPI validation schema constructs to restrict `"x"` further.
+
+Note that we lose expressiveness of the existing `format` strings: we either apply the format to `"json"` or any other pre-defined format. I.e. we cannot express that pruning should be disabled, but if it is an integer, it should be an int32.
+
+**Question:** this feels like a reasonable loss of expressivity. Do we accept that?
+
+**Raw JSON Option 2:** add an extension property, e.g.
+
+```json
+{"properties": {"x": {"x-no-pruning": true}}}
```

Comment:
**Mehdy Bohlool:** While I think we don't need first approach and map[string]object is expressing that, I like to have extension. There could be cases that I want to express some validations on some of the fields of a json object but I don't want pruning of other fields in that sub-object. e.g. is when the sub-object has ObjectMeta or TypeMeta but the rest of the object is a free json.